{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "51KZoXVXELF9",
        "outputId": "9be06ce8-93d3-4f83-9eb7-d7c77ef5cf4b"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n",
            "/content/drive/.shortcut-targets-by-id/1dFGZRZcbpBvtZJUDvrdI5mVkcMo0IDm5/GC Merge Data\n",
            " \u001b[0m\u001b[01;36mdata\u001b[0m@             \u001b[01;36m'GC-MERGE example'\u001b[0m@       process_inputs_.py\n",
            "'Data Test.ipynb'  'Processing Data.ipynb'   \u001b[01;34m__pycache__\u001b[0m/\n"
          ]
        }
      ],
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive', force_remount=True)\n",
        "\n",
        "%cd \"drive/MyDrive/GC Merge Data\"\n",
        "%ls"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "_JjGY96mEX2A"
      },
      "outputs": [],
      "source": [
        "import os\n",
        "import numpy as np\n",
        "import scipy.stats as scp\n",
        "from scipy.sparse import csr_matrix\n",
        "from scipy.sparse import save_npz\n",
        "\n",
        "import pandas as pd\n",
        "\n",
        "import pickle\n",
        "from types import SimpleNamespace"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "NUC3LTAJE8Mq"
      },
      "outputs": [],
      "source": [
        "# %load process_inputs_.py\n",
        "# from process_inputs_ import *"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pwd"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 36
        },
        "id": "jbknv3NNgeXn",
        "outputId": "707bf5b3-022b-448e-d8cd-e41724fa1140"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'/content/drive/.shortcut-targets-by-id/1dFGZRZcbpBvtZJUDvrdI5mVkcMo0IDm5/GC Merge Data'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 4
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Imports"
      ],
      "metadata": {
        "id": "SaoXDLuPcYd-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def gen_hic_map(hic_matrix_coord, weights_counts, hic_res, kth):\n",
        "    '''\n",
        "    Generates Hi-C matrix for each chromosome\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    hic_matrix_coord [array]: 2-column array of Hi-C bin coordinates (i, j)\n",
        "    weights_counts [array]: 1-columnn array of bin counts corresponding to (i, j)\n",
        "    hic_res [int]: Hi-C map resolution in base pairs (10000 bp)\n",
        "    kth [int]: Number of neighbors\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    hic_map_reset [array]: Square Hi-C matrix for each chromosome (rows, columns start at zero)\n",
        "    edge_wts [array]: Bin counts corresponding to non-zero entries\n",
        "    row_idx_vec [array]: Row indices of non-zero matrix entries\n",
        "    col_idx_vec [array]: Column indices of non-zero matrix entries\n",
        "    hic_map_reset_range [int]: Number of rows (or equivalently columns) in Hi-C matrix\n",
        "    chrom_start_local [int]: Minimum chromosome coordinate for chromosome's\n",
        "        Hi-C map (corresponds to local start)\n",
        "\n",
        "    '''\n",
        "\n",
        "    chrom_start_local = np.min(hic_matrix_coord[:, 0])\n",
        "    num_bins_sq = int((np.max(hic_matrix_coord[:, 1]) - np.min(hic_matrix_coord[:, 0]))/hic_res)\n",
        "    chrom_finish_local = chrom_start_local + num_bins_sq*hic_res\n",
        "\n",
        "    bin_edges_x = np.arange(chrom_start_local, chrom_finish_local + hic_res, hic_res)\n",
        "    bin_edges_y = np.arange(chrom_start_local, chrom_finish_local + hic_res, hic_res)\n",
        "\n",
        "    hic_map_reset, xedges, yedges = np.histogram2d(hic_matrix_coord[:,0], hic_matrix_coord[:,1], \\\n",
        "        weights = weights_counts, bins = [bin_edges_x, bin_edges_y])\n",
        "\n",
        "    hic_map_reset_range = np.size(hic_map_reset, 0)\n",
        "\n",
        "    ### Reflects matrix over diagonal to ensure symmetry\n",
        "    idx_upper = np.tril_indices(num_bins_sq, k=-1)\n",
        "    hic_map_reset[idx_upper] = hic_map_reset.T[idx_upper]\n",
        "\n",
        "    row_idx_mat, col_idx_mat = np.indices(np.shape(hic_map_reset))\n",
        "\n",
        "    ### Subtracts medians of upper diagonals for each entry and sets negative entries to zero\n",
        "    nz_k_list = [k for k in range(hic_map_reset_range) if np.any(np.diagonal(hic_map_reset,k) > 0)]\n",
        "    for k in nz_k_list:\n",
        "        diag_k = np.diagonal(hic_map_reset, k)\n",
        "        diag_median = np.median(diag_k)\n",
        "        row_vals = np.diag(row_idx_mat, k)\n",
        "        col_vals = np.diag(col_idx_mat, k)\n",
        "        hic_map_reset[row_vals, col_vals] = hic_map_reset[row_vals, col_vals] - diag_median\n",
        "    hic_map_reset[hic_map_reset < 0] = 0\n",
        "\n",
        "    ### Reflects matrix over diagonal to ensure symmetry\n",
        "    idx_upper = np.tril_indices(num_bins_sq, k=-1)\n",
        "    hic_map_reset[idx_upper] = hic_map_reset.T[idx_upper]\n",
        "\n",
        "    ### Sets diagonals = 0 to ensure only neighbor nodes are selected\n",
        "    hic_map_reset[np.diag_indices(np.size(hic_map_reset, 0))] = 0\n",
        "\n",
        "    \"\"\" OLD, WE DONT WANT TO SUBSAMPLE\n",
        "    # ### Subsamples matrix by taking only the top kth neighbors for each row\n",
        "    # max_row, max_col = np.shape(hic_map_reset)\n",
        "    # top_kth_col_idcs_mat = np.argpartition(hic_map_reset, -kth, axis=-1)[:,-kth:]\n",
        "    # top_kth_col_idcs_vec = top_kth_col_idcs_mat.flatten()\n",
        "    # top_kth_rows_mat = np.reshape(np.arange(max_row), (-1,1))\n",
        "    # top_kth_rows_mat = np.repeat(top_kth_rows_mat, kth, axis=-1)\n",
        "    # top_kth_rows_vec = top_kth_rows_mat.flatten()\n",
        "    # top_kth_vals_vec = hic_map_reset[top_kth_rows_vec, top_kth_col_idcs_vec]\n",
        "\n",
        "    # hic_map_reset_filtered = np.zeros(np.shape(hic_map_reset))\n",
        "    # hic_map_reset_filtered[top_kth_rows_vec, top_kth_col_idcs_vec] = top_kth_vals_vec\n",
        "\n",
        "    # ### If count(i, j) != count(j, i), takes the higher of the two counts\n",
        "    # for (i, j) in list(zip(top_kth_rows_vec, top_kth_col_idcs_vec)):\n",
        "    \"\"\"\n",
        "\n",
        "    hic_map_reset_filtered = hic_map_reset\n",
        "    row_idx_vec = row_idx_mat[np.nonzero(hic_map_reset_filtered)].flatten()\n",
        "    col_idx_vec = col_idx_mat[np.nonzero(hic_map_reset_filtered)].flatten()\n",
        "\n",
        "    for (i,j) in list(zip(row_idx_vec, col_idx_vec)):\n",
        "        max_val = np.max((hic_map_reset_filtered[i, j], hic_map_reset_filtered[j, i]))\n",
        "        hic_map_reset_filtered[i, j] = hic_map_reset_filtered[j, i] = max_val\n",
        "\n",
        "    ### Reflects matrix over diagonal to ensure symmetry\n",
        "    idx_upper = np.tril_indices(hic_map_reset_range, k=-1)\n",
        "    hic_map_reset_filtered[idx_upper] = hic_map_reset_filtered.T[idx_upper]\n",
        "    hic_map_reset_filtered[np.diag_indices(np.size(hic_map_reset_filtered, 0))] = 0\n",
        "\n",
        "    ### Stores row indices, column indices, and bin counts to construct sparse csr matrix\n",
        "    # row_idx_vec = row_idx_mat[np.nonzero(hic_map_reset_filtered)].flatten()\n",
        "    # col_idx_vec = col_idx_mat[np.nonzero(hic_map_reset_filtered)].flatten()\n",
        "    edge_wts = hic_map_reset_filtered[np.nonzero(hic_map_reset_filtered)].flatten()\n",
        "\n",
        "    hic_map_reset = hic_map_reset_filtered\n",
        "\n",
        "    return hic_map_reset, edge_wts, row_idx_vec, col_idx_vec, \\\n",
        "        hic_map_reset_range, chrom_start_local\n",
        "\n",
        "\n",
        "def get_adj(hic_map_reset, hic_node_start, kth):\n",
        "    '''\n",
        "    Generates dictionary of nodes and neighbors as well as whether the node\n",
        "        is connected or disconnected\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    hic_map_reset [array]: Square Hi-C matrix for each chromosome\n",
        "    hic_node_start [int]: New starting coordinate for chromosome\n",
        "        concatenated Hi-C matrix\n",
        "    kth [int]: Number of neighbors\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    adj_binary_reset [arr]: 1-column array of 0s (disconnected node)\n",
        "        and 1s (connected node)\n",
        "\n",
        "    '''\n",
        "\n",
        "    num_nodes_reset = np.size(hic_map_reset, 0)\n",
        "    adj_binary_reset = np.zeros(num_nodes_reset)\n",
        "\n",
        "    for i in list(np.arange(0, num_nodes_reset, 1)):\n",
        "\n",
        "        row = hic_map_reset[i, :]\n",
        "\n",
        "        if np.any(row > 0) == True:\n",
        "            adj_binary_reset[i] = 1\n",
        "\n",
        "    return adj_binary_reset\n",
        "\n",
        "\n",
        "def get_hmods(cell_line_dir, cell_line, chr_num, hic_res, chip_res, hic_node_start, chrom_start_local, hic_map_reset_range):\n",
        "    '''\n",
        "    Generates array of nodes (rows) and corresponding histone modification\n",
        "        features (columns)\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    cell_line_dir [str]: Path to data directory\n",
        "    cell_line [str]: Name of cell line\n",
        "    chr_num [int]: Chromosome number\n",
        "    hic_res [int]: Hi-C map resolution in base pairs (10000 bp)\n",
        "    chip_res [int]: ChIP-seq resolution in base pairs (10000 bp)\n",
        "\n",
        "    hic_node_start [int]: New starting coordinate for chromosome\n",
        "        concatenated Hi-C matrix\n",
        "    chrom_start_local [int]: Minimum chromosome coordinate for chromosome's\n",
        "        Hi-C map (corresponds to local start)\n",
        "    hic_map_reset_range [int]: Number of rows (or equivalently columns) in Hi-C matrix\n",
        "\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    np_hmods_norm [array]: Array of node features (histone modification signals\n",
        "        per node)\n",
        "    col_name_list [str list]: List of feature names (histone modification types)\n",
        "\n",
        "    '''\n",
        "\n",
        "    os.chdir(cell_line_dir)\n",
        "    dir1_encode = os.fsencode(cell_line_dir)\n",
        "\n",
        "    h_mod_norm_list = []\n",
        "    col_name_list = []\n",
        "\n",
        "    num_chip_feat = hic_res/chip_res\n",
        "\n",
        "    for file in sorted(os.listdir(dir1_encode)):\n",
        "\n",
        "         filename = os.fsdecode(file)\n",
        "         #print(filename)\n",
        "\n",
        "#         if filename.startswith('chr' + str(chr_num) + '_' + str(chip_res) + 'bp_') and filename.endswith('.count'):\n",
        "#         if filename.startswith('chr' + str(chr_num) + '_' + str(chip_res) + 'bp_') and filename.endswith('.count') and (not ('DNase' in filename)):\n",
        "         if filename.startswith('chr' + str(chr_num) + '_') and filename.endswith('.count'): #and (not ('DNase' in filename)):\n",
        "#         if filename.startswith('chr' + str(chr_num) + '_' + str(chip_res) + 'bp_') and filename.endswith('.count') and (not ('H3K27ac' in filename)) and (not ('DNase' in filename)):\n",
        "\n",
        "             col_name = filename.split('_')[1]\n",
        "             print(filename)\n",
        "             #changed here\n",
        "\n",
        "             col_name = col_name.split('.')[0]\n",
        "             col_name_list.append(col_name)\n",
        "\n",
        "             ### Columns = start, finish, count\n",
        "             h_mod = np.loadtxt(filename, delimiter='\\t', usecols=[3])\n",
        "\n",
        "             ### Normalizes histone modification signal values relative to maximum\n",
        "             h_mod_norm = h_mod/np.max(h_mod)\n",
        "\n",
        "             num_rows_total = np.size(h_mod_norm, 0)\n",
        "             num_rows_reduced = int(num_rows_total//num_chip_feat)\n",
        "             h_mod_norm_mat = np.reshape(h_mod_norm, (num_rows_reduced, -1))\n",
        "             h_mod_norm_list.append(h_mod_norm_mat)\n",
        "\n",
        "\n",
        "    print(\"Test\")\n",
        "\n",
        "\n",
        "    #print(h_mod_norm_list)\n",
        "    #print(np_hmods_chr)\n",
        "\n",
        "    temp = [x.shape for x in h_mod_norm_list]\n",
        "    print(temp)\n",
        "\n",
        "    np_hmods_chr = np.concatenate((h_mod_norm_list), axis=1)\n",
        "\n",
        "\n",
        "    node_id_seq_local = np.reshape(np.arange(num_rows_reduced), (-1, 1))\n",
        "    node_id_seq = np.floor(node_id_seq_local + hic_node_start).astype(int)\n",
        "\n",
        "    np_hmods_norm_pre = np.concatenate((node_id_seq, np_hmods_chr), axis=1)\n",
        "\n",
        "    np_hmods_norm = np_hmods_norm_pre[:hic_map_reset_range, :]\n",
        "\n",
        "    return np_hmods_norm, col_name_list\n",
        "\n",
        "\n",
        "def get_labels(rnaseq_file, cell_line, regression_flag):\n",
        "    '''\n",
        "    Generates dataframe containing nodes and true target labels\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    rnaseq_file [str]: Path to file containing RNA-seq measurements\n",
        "    cell_line [str]: Name of cell line\n",
        "    regression_flag [int]: Task formulation type, where 1 = regression\n",
        "        and 0 = classification\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    df_labels [dataframe]: Dataframe containing true node labels\n",
        "\n",
        "    '''\n",
        "\n",
        "    if type(cell_line) != str:\n",
        "        cell_line = str(cell_line)\n",
        "    df_rnaseq = pd.read_csv(rnaseq_file)\n",
        "\n",
        "    df_rnaseq = df_rnaseq[['gene_id', cell_line]]\n",
        "\n",
        "    ### For classification, binarize expression levels relative to median\n",
        "    ### For regression, add pseudocount and then take log10, s.t. all values >= 0\n",
        "\n",
        "    if regression_flag == 0:\n",
        "        med = df_rnaseq[cell_line].median()\n",
        "        df_rnaseq[cell_line] = np.where(df_rnaseq[cell_line] > med, 1, 0)\n",
        "    else:\n",
        "        df_rnaseq[cell_line] = np.log10(df_rnaseq[cell_line] + 1)\n",
        "\n",
        "    df_labels = df_rnaseq.rename(columns={\"gene_id\": \"gene_catalog_name\"})\n",
        "    df_labels = df_labels.rename(columns={cell_line: \"expression_lvl\"})\n",
        "\n",
        "    return df_labels\n",
        "\n",
        "\n",
        "def get_gene_coord(gene_coord_file, chr_num, hic_res, hic_node_start, chrom_start_local):\n",
        "    '''\n",
        "    Generates dataframe of gene coordinates\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    gene_coord_file [str]: Path to gene coordinate file\n",
        "    chr_num [int]: Chromosome number\n",
        "    hic_res [int]: Hi-C map resolution in base pairs (10000 bp)\n",
        "    hic_node_start [int]: New starting coordinate for chromosome\n",
        "        concatenated Hi-C matrix\n",
        "    chrom_start_local [int]: Minimum chromosome coordinate for chromosome's\n",
        "        Hi-C map (corresponds to local start)\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    df_gene_coord [dataframe]: Dataframe of gene coordinates\n",
        "\n",
        "    '''\n",
        "    df = pd.read_csv(gene_coord_file, sep=',|\\s+|\\t+', engine='python', header=None, \\\n",
        "                      names=['gene_catalog_name', 'Chr', 'Start', 'Finish', 'Polarity', 'Category', 'Abbrev', 'Desc'])\n",
        "\n",
        "    df_gene_coord = df.loc[df['Chr'] == str(chr_num)]\n",
        "\n",
        "    df_gene_coord['TSS'] = np.where(df_gene_coord['Polarity'] == 1, df_gene_coord['Start'], df_gene_coord['Finish'])\n",
        "\n",
        "\n",
        "    df_gene_coord['hic_node_id'] = np.floor(hic_node_start +\n",
        "                  np.divide(df_gene_coord['TSS'].to_numpy() - chrom_start_local, hic_res))\n",
        "    df_gene_coord['hic_node_id'] = df_gene_coord['hic_node_id'].astype('int32', copy=False)\n",
        "\n",
        "    # df_gene_coord['hic_node_id'] = np.floor(hic_node_start +\n",
        "    #              np.divide(df_gene_coord['TSS'].to_numpy() - chrom_start_local, hic_res)).astype(int)\n",
        "\n",
        "    df_gene_coord = df_gene_coord[df_gene_coord['hic_node_id'] >= 0]\n",
        "\n",
        "    df_gene_coord = df_gene_coord.sort_values(by=['hic_node_id'])\n",
        "\n",
        "    return df_gene_coord\n",
        "\n",
        "\n",
        "def remove_dup(df_gene_coord, df_labels, regression_flag):\n",
        "    '''\n",
        "    For each node containing more than one gene, if regression_flag = 0,\n",
        "        i.e. binary classification task, take the mode of the binarized\n",
        "        expression values as the node label; else if regression_flag = 1,\n",
        "        i.e. continuous regression task, take the median of the expression\n",
        "        values and assign this as the node label; in both cases, return a\n",
        "        dictionary with keys corresponding to all nodes with more than one\n",
        "        gene and values corresponding to [gene names, gene expression values,\n",
        "        central tendency (mode or median)]\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    df_gene_coord [dataframe]: Dataframe of gene coordinates\n",
        "    df_labels [dataframe]: Dataframe containing true node labels\n",
        "    regression_flag [int]: Task formulation type, where 1 = regression\n",
        "        and 0 = classification\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    df_genes_nodup [dataframe]: Dataframe of gene coordinates with duplicate\n",
        "        expression values removed\n",
        "    duplicate_dict [dict]: Dictionary of {node: [gene names, gene expression\n",
        "        values, central tendency]} for all nodes that correspond to more than one gene\n",
        "\n",
        "    '''\n",
        "    grp_exp_val_list = []\n",
        "    grp_central_tendency_list = []\n",
        "    grp_gene_id_list = []\n",
        "    grp_hic_id_list = []\n",
        "    duplicate_dict = {}\n",
        "\n",
        "    df_gene_lab = pd.merge(df_gene_coord, df_labels, on='gene_catalog_name', how='inner')\n",
        "\n",
        "    for hic_id, group in df_gene_lab.groupby(by=\"hic_node_id\"):\n",
        "        if len(group) > 1:\n",
        "            grp_hic_id_list.append(hic_id)\n",
        "            grp_exp_vals = group[\"expression_lvl\"].values.tolist()\n",
        "            grp_exp_val_list.append(grp_exp_vals)\n",
        "\n",
        "            grp_gene_names = group[\"gene_catalog_name\"].tolist()\n",
        "            grp_gene_id_list.append(grp_gene_names)\n",
        "\n",
        "            if regression_flag == 0:\n",
        "                grp_central_tendency = int(scp.mode(grp_exp_vals)[0])\n",
        "                grp_central_tendency_list.append(grp_central_tendency)\n",
        "            else:\n",
        "                grp_central_tendency = np.median(grp_exp_vals)\n",
        "                grp_central_tendency_list.append(grp_central_tendency)\n",
        "\n",
        "            duplicate_dict[hic_id] = [grp_gene_names, grp_exp_vals, grp_central_tendency]\n",
        "\n",
        "    df_genes_nodup = df_gene_lab.drop_duplicates(subset=\"hic_node_id\", keep=\"first\")\n",
        "    df_genes_nodup = df_genes_nodup.sort_values(by=['hic_node_id'])\n",
        "\n",
        "    for h in range(len(grp_hic_id_list)):\n",
        "        hic_id = grp_hic_id_list[h]\n",
        "        df_genes_nodup.loc[df_genes_nodup[\"hic_node_id\"] == hic_id,\\\n",
        "                                        \"expression_lvl\"] = grp_central_tendency_list[h]\n",
        "\n",
        "    return df_genes_nodup, duplicate_dict\n",
        "\n",
        "\n",
        "def merge_dfs_inner(df_genes_nodup, adj_binary_reset, hic_node_start):\n",
        "    '''\n",
        "    Generates dataframe of nodes with genes and their labels\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    df_genes_nodup [dataframe]: Dataframe of gene coordinates with duplicate\n",
        "        expression values removed\n",
        "    adj_binary_reset [arr]: 1-column array of 0s (disconnected node)\n",
        "        and 1s (connected node)\n",
        "    hic_node_start [int]: New starting coordinate for chromosome\n",
        "        concatenated Hi-C matrix\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    df_genes_lab_nodup_connected [dataframe]: Dataframe of connected nodes\n",
        "        with genes and their corresponding expression labels\n",
        "\n",
        "    '''\n",
        "\n",
        "    df_adj_binary_reset = pd.DataFrame()\n",
        "\n",
        "    df_adj_binary_reset['hic_node_id'] = np.arange(hic_node_start, hic_node_start + np.size(adj_binary_reset), 1)\n",
        "    df_adj_binary_reset['connected'] = adj_binary_reset\n",
        "\n",
        "    df_genes_lab_nodup  = pd.merge(df_genes_nodup, df_adj_binary_reset, on='hic_node_id', how='inner')\n",
        "\n",
        "    #for nodes with no genes, set label = -1 (though technically this should be taken care of by inner merge)\n",
        "    df_genes_lab_nodup['expression_lvl'] = np.where(np.isnan(df_genes_lab_nodup['expression_lvl'])==True, \\\n",
        "                          -1, df_genes_lab_nodup['expression_lvl'])\n",
        "\n",
        "    #double-confirmation that genes with no nodes are not included\n",
        "    df_genes_lab_nodup = df_genes_lab_nodup.loc[df_genes_lab_nodup['expression_lvl'] != -1]\n",
        "\n",
        "    #for nodes with no neighbors, remove these nodes from gene prediction df\n",
        "    df_genes_lab_nodup_connected = df_genes_lab_nodup.loc[df_genes_lab_nodup['connected'] == 1]\n",
        "    df_genes_lab_nodup_connected = df_genes_lab_nodup_connected.sort_values(by=['hic_node_id'])\n",
        "\n",
        "    return df_genes_lab_nodup_connected\n",
        "\n",
        "\n",
        "def merge_dfs_outer(np_hmods_norm, df_genes_nodup, chr_num, chrom_start_local, hic_res):\n",
        "    '''\n",
        "    Generates merged dataframe with all nodes and their chromosome coordinates\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    np_hmods_norm [array]: Array of node features (histone modification signals\n",
        "        per node)\n",
        "    df_genes_nodup [dataframe]: Dataframe of gene coordinates with duplicate\n",
        "        expression values removed\n",
        "    chr_num [int]: Chromosome number\n",
        "    chrom_start_local [int]: Minimum chromosome coordinate for chromosome's\n",
        "        Hi-C map (corresponds to local start)\n",
        "    hic_res [int]: Hi-C map resolution in base pairs (10000 bp)\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    df_node_coord [dataframe]: Merged dataframe containing all nodes and their\n",
        "        chromosome coordinates\n",
        "    '''\n",
        "\n",
        "    num_nodes = np.size(np_hmods_norm, 0)\n",
        "    hic_node_id_vec = np.reshape(np_hmods_norm[:,0], (-1, 1))\n",
        "    chr_start_vec = np.arange(chrom_start_local, chrom_start_local + num_nodes*hic_res, hic_res)\n",
        "    chr_start_vec = np.reshape(chr_start_vec, (num_nodes, 1))\n",
        "    chr_finish_vec = chr_start_vec + hic_res\n",
        "    chr_num_vec = np.ones((num_nodes, 1))*chr_num\n",
        "\n",
        "    np_coord = np.concatenate((hic_node_id_vec, chr_num_vec, chr_start_vec, chr_finish_vec), axis = 1)\n",
        "    df_node_coord = pd.DataFrame(np_coord, columns=['hic_node_id', 'chr_num', 'chr_start', 'chr_finish'])\n",
        "\n",
        "    df_node_coord  = pd.merge(df_node_coord, df_genes_nodup, on='hic_node_id', how='outer')\n",
        "\n",
        "    return df_node_coord\n",
        "\n",
        "\n",
        "def process_chrom(obs_file, rnaseq_file, gene_coord_file, cell_line_dir, cell_line,\n",
        "    chr_num, hic_res, chip_res, kth, hic_node_start, regression_flag):\n",
        "    '''\n",
        "    Processes each chromosome\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    obs_file: Tab-delimited Hi-C file (.txt) containing three columns:\n",
        "        [bin coordinate 1, bin coordinate 2, Hi-C count]\n",
        "    rnaseq_file [array]: Path to file containing RNA-seq measurements\n",
        "    gene_coord_file [str]: Path to gene coordinate file\n",
        "    cell_line_dir [str]: Path to data directory\n",
        "    cell_line [str]: Name of cell line\n",
        "    chr_num [int]: Chromosome number\n",
        "    hic_res [int]: Hi-C map resolution in base pairs (10000 bp)\n",
        "    chip_res [int]: ChIP-seq resolution in base pairs (10000 bp)\n",
        "    kth [int]: Number of neighbors\n",
        "    hic_node_start [int]: New starting coordinate for chromosome\n",
        "        concatenated Hi-C matrix\n",
        "    regression_flag [int]: Task formulation type, where 1 = regression\n",
        "        and 0 = classification\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    np_genes_lab_cat [array]: Array of genes and labels for each chromosome\n",
        "    np_hmods_norm [array]: Array of node features (histone modification signals\n",
        "        per node)\n",
        "    df_node_coord [dataframe]: Merged dataframe containing all nodes and their\n",
        "        chromosome coordinates\n",
        "    hic_map_reset_range [int]: Number of rows (or equivalently columns) in Hi-C matrix\n",
        "    edge_wts [array]: Bin counts corresponding to non-zero entries\n",
        "    row_idx_vec [array]: Row indices of non-zero matrix entries\n",
        "    col_idx_vec [array]: Column indices of non-zero matrix entries\n",
        "    df_genes_lab_nodup_connected_names : Dataframe of connected nodes with\n",
        "        genes, genes names, and expression labels\n",
        "    duplicate_dict [dict]: Dictionary of {node: [gene names, gene expression\n",
        "        values, gene expression mode]} for all nodes that correspond to\n",
        "        more than one gene\n",
        "\n",
        "    '''\n",
        "\n",
        "    print('chr:', chr_num)\n",
        "\n",
        "    obs_mat = np.loadtxt(obs_file)\n",
        "\n",
        "    if np.size(obs_mat, 1) > np.size(obs_mat, 0):\n",
        "        obs_mat = obs_mat.T\n",
        "\n",
        "    obs_mat[np.isnan(obs_mat[:,2]), 2] = 0\n",
        "\n",
        "    hic_matrix_coord = obs_mat[:,:2]\n",
        "\n",
        "    weights_counts = obs_mat[:, 2]\n",
        "\n",
        "    chrom_start_local = np.min(hic_matrix_coord[:, 0])\n",
        "\n",
        "    df_labels = get_labels(rnaseq_file, cell_line, regression_flag)\n",
        "\n",
        "    df_gene_coord = get_gene_coord(gene_coord_file, chr_num, hic_res, hic_node_start, chrom_start_local)\n",
        "    df_gene_coord = df_gene_coord.drop(columns=['Start', 'Finish', 'Polarity', \\\n",
        "                    'Category', 'Desc', 'TSS', 'Chr'])\n",
        "\n",
        "    df_genes_nodup, duplicate_dict = remove_dup(df_gene_coord, df_labels, regression_flag)\n",
        "\n",
        "    hic_map_reset, edge_wts, row_idx_vec, col_idx_vec, hic_map_reset_range, chrom_start_local = \\\n",
        "        gen_hic_map(hic_matrix_coord, weights_counts, hic_res, kth)\n",
        "\n",
        "    adj_binary_reset = get_adj(hic_map_reset, hic_node_start, kth)\n",
        "\n",
        "    np_hmods_norm, col_name_list = get_hmods(cell_line_dir, cell_line, chr_num, hic_res, chip_res, hic_node_start, chrom_start_local, hic_map_reset_range)\n",
        "\n",
        "    df_genes_lab_nodup_connected_names = merge_dfs_inner(df_genes_nodup, adj_binary_reset, hic_node_start)\n",
        "\n",
        "    df_genes_lab_filtered = df_genes_lab_nodup_connected_names.drop(columns=['connected', 'Abbrev', 'gene_catalog_name'])\n",
        "\n",
        "    np_genes_lab_filtered = df_genes_lab_filtered.to_numpy()\n",
        "\n",
        "    #only predict for nodes with genes\n",
        "    #note that the embedding matrix must still use all of the nodes though!\n",
        "    np_genes_lab_cat = np_genes_lab_filtered[np_genes_lab_filtered[:,-2] != -1]\n",
        "    #now we have restricted np_nodes_lab to only connected nodes with genes\n",
        "\n",
        "    df_node_coord = merge_dfs_outer(np_hmods_norm, df_genes_nodup, chr_num, chrom_start_local, hic_res)\n",
        "\n",
        "    return np_genes_lab_cat, np_hmods_norm, df_node_coord, hic_map_reset_range, \\\n",
        "        edge_wts, row_idx_vec, col_idx_vec, df_genes_lab_nodup_connected_names, duplicate_dict"
      ],
      "metadata": {
        "id": "4V88umRvcSp7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Main"
      ],
      "metadata": {
        "id": "aZ78qqD3ccnm"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "args = SimpleNamespace()\n",
        "args.cell_line ='E116'\n",
        "args.num_neighbors = 10\n",
        "args.chip_resolution = 10000\n",
        "args.regression_flag = 0\n",
        "args.version ='_redone_h3k27ac_addition'\n",
        "args.save_file_flags ='hcr'\n",
        "\n",
        "\n",
        "# save_file_flags can be any combination of 'h' (save Hi-C matrix),\n",
        "#   'c' (save all ChIP-seq resolution dependent files), and\n",
        "#   'r' (save all regression-dependent files)\n",
        "\n",
        "cell_line = args.cell_line\n",
        "kth = args.num_neighbors\n",
        "chip_res = args.chip_resolution\n",
        "regression_flag = args.regression_flag\n",
        "version = args.version\n",
        "save_file_flags = args.save_file_flags\n",
        "\n",
        "np.random.seed(1)\n",
        "hic_res = 10000\n",
        "num_hm = 6\n",
        "\n",
        "base_path = \"/content/drive/Shareddrives/DeepBio/GC Merge Data/\"\n",
        "rnaseq_file = os.path.join(base_path, 'data', 'gc_merge_processed_data', 'rnaseq.csv')\n",
        "gene_coord_file = os.path.join(base_path, 'data', 'gc_merge_processed_data', 'gene_coord.csv')\n",
        "cell_line_dir = os.path.join(base_path , 'GC-MERGE example')\n",
        "save_dir = os.path.join(base_path , 'data', args.cell_line, f\"{hic_res}\")\n",
        "\n",
        "\n",
        "chr_num_low_inc = 13\n",
        "chr_num_high_exc = 14\n",
        "\n",
        "\n",
        "\n",
        "### Initializes blank data structures\n",
        "num_feat_bins = int(hic_res/chip_res)\n",
        "np_hmods_norm_all = np.zeros((1, num_hm*num_feat_bins))\n",
        "\n",
        "\n",
        "\n",
        "for i in range(chr_num_low_inc, chr_num_high_exc):\n",
        "    # Initialize arrays for each chromosome to reset data\n",
        "    np_genes_lab = np.zeros((1, 2))\n",
        "    np_hmods_norm_all = np.zeros((1, num_hm*num_feat_bins))\n",
        "    riv_list = []\n",
        "    civ_list = []\n",
        "    edge_wts_list = []\n",
        "    df_gene_names_ids = []\n",
        "    df_node_coord_list = []\n",
        "    duplicate_dict_all = {}\n",
        "\n",
        "    hic_node_start = 0  # Reset hic_node_start for each chromosome\n",
        "\n",
        "    chr_num = i\n",
        "    print('Processing chromosome ' + str(chr_num))\n",
        "    obs_file = os.path.join(cell_line_dir, 'hic_chr' + str(chr_num) + '.txt')\n",
        "\n",
        "    # Process chromosome\n",
        "    np_genes_lab_cat, np_hmods_norm, df_node_coord, hic_map_reset_range, \\\n",
        "    edge_wts, riv, civ, df_genes_lab_nodup_connected_names_cat, duplicate_dict = \\\n",
        "    process_chrom(obs_file, rnaseq_file, gene_coord_file, cell_line_dir, cell_line, \\\n",
        "                  chr_num, hic_res, chip_res, kth, hic_node_start, regression_flag)\n",
        "\n",
        "    # Concatenate and process data for the current chromosome\n",
        "    np_genes_lab = np.concatenate((np_genes_lab, np_genes_lab_cat), axis=0)\n",
        "\n",
        "    # Concatenate histone modification data\n",
        "    np_hmods_norm_all = np.concatenate((np_hmods_norm_all, np_hmods_norm), axis=0)\n",
        "\n",
        "    df_gene_names_ids.append(df_genes_lab_nodup_connected_names_cat)\n",
        "    df_node_coord_list.append(df_node_coord)\n",
        "\n",
        "    # Adjust indices based on hic_node_start\n",
        "    riv = riv + hic_node_start\n",
        "    civ = civ + hic_node_start\n",
        "    riv_list.extend(riv)\n",
        "    civ_list.extend(civ)\n",
        "    edge_wts_list.extend(edge_wts)\n",
        "\n",
        "    hic_node_start += hic_map_reset_range\n",
        "    duplicate_dict_all.update(duplicate_dict)\n",
        "\n",
        "    # Finalize and save processed data for the current chromosome\n",
        "    np_genes_lab = np.delete(np_genes_lab, 0, axis=0)\n",
        "    np_hmods_norm_all = np.delete(np_hmods_norm_all, 0, axis=0)\n",
        "\n",
        "    # Finalizes gene-feature-label dataframe\n",
        "    mask = np.isin(np_hmods_norm_all[:,0], np_genes_lab[:,0])\n",
        "    np_genes_hmods_lab = np_hmods_norm_all[mask, :]\n",
        "    # print(f\"np_genes_hmods_lab: {np_genes_hmods_lab.shape}\")\n",
        "    # print(f\"np_genes_lab: {np_genes_lab.shape}\")\n",
        "    np_genes_hmods_lab = np.concatenate((np_genes_hmods_lab, np.reshape(np_genes_lab[:,-1], (-1, 1))), axis=1)\n",
        "\n",
        "    # Generates sparse csr matrix\n",
        "    edge_wts_vec = np.array(edge_wts_list)\n",
        "    riv_vec = np.array(riv_list)\n",
        "    civ_vec = np.array(civ_list)\n",
        "\n",
        "    print(riv_vec.shape)\n",
        "    print(civ_vec.shape)\n",
        "    print(hic_node_start)\n",
        "    hic_sparse_mat = csr_matrix((edge_wts_vec, (riv_vec, civ_vec)), shape=(hic_node_start, hic_node_start))\n",
        "\n",
        "    df_genes = pd.concat(df_gene_names_ids, axis=0)\n",
        "    df_genes.rename(columns={\"Abbrev\": \"abbrev\"}, inplace=True)\n",
        "\n",
        "\n",
        "\n",
        "    ### Saves processed files\n",
        "    for flag in save_file_flags:\n",
        "      print(flag)\n",
        "      if flag == 'h':\n",
        "        hic_sparse_mat_file = os.path.join(save_dir, f'chr{chr_num}', 'hic_sparse' + version + '.npz')\n",
        "        print(hic_sparse_mat_file)\n",
        "        save_npz(hic_sparse_mat_file, hic_sparse_mat)\n",
        "\n",
        "      elif flag == 'c':\n",
        "\n",
        "        np_hmods_norm_all_file = os.path.join(save_dir, f'chr{chr_num}', 'np_hmods_norm_chip_' + \\\n",
        "            str(chip_res) + 'bp' + version + '.npy')\n",
        "        np.save(np_hmods_norm_all_file, np_hmods_norm_all, allow_pickle=True)\n",
        "\n",
        "        # df_node_coord_file = os.path.join(cell_line_dir,  \\\n",
        "        #                                     'df_node_coord' + version + '.pkl')\n",
        "        # df_node_coord.to_pickle(df_node_coord_file)\n",
        "\n",
        "      elif flag == 'r':\n",
        "        np_genes_lab_file = os.path.join(save_dir, f'chr{chr_num}', 'np_nodes_lab_genes_reg' + \\\n",
        "            str(regression_flag) + version + '.npy')\n",
        "        np.save(np_genes_lab_file, np_genes_lab, allow_pickle=True)\n",
        "\n",
        "        df_genes_file = os.path.join(save_dir, f'chr{chr_num}', 'df_genes_reg' + str(regression_flag) + version + '.pkl')\n",
        "        df_genes.to_pickle(df_genes_file)\n",
        "\n",
        "        dict_gene_dups_file = os.path.join(save_dir, f'chr{chr_num}', 'dict_gene_dups_reg' + \\\n",
        "            str(regression_flag) + version + '.pkl')\n",
        "        dict_gene_dups = open(dict_gene_dups_file,\"wb\")\n",
        "        pickle.dump(duplicate_dict_all, dict_gene_dups)\n",
        "        dict_gene_dups.close()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fBg3rTGTXH6Z",
        "outputId": "4b2e896d-9aa9-436e-a99f-e22f5f459716"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Processing chromosome 13\n",
            "chr: 13\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-3-513c0ed21cb9>:281: SettingWithCopyWarning: \n",
            "A value is trying to be set on a copy of a slice from a DataFrame.\n",
            "Try using .loc[row_indexer,col_indexer] = value instead\n",
            "\n",
            "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
            "  df_gene_coord['TSS'] = np.where(df_gene_coord['Polarity'] == 1, df_gene_coord['Start'], df_gene_coord['Finish'])\n",
            "<ipython-input-3-513c0ed21cb9>:284: SettingWithCopyWarning: \n",
            "A value is trying to be set on a copy of a slice from a DataFrame.\n",
            "Try using .loc[row_indexer,col_indexer] = value instead\n",
            "\n",
            "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
            "  df_gene_coord['hic_node_id'] = np.floor(hic_node_start +\n",
            "<ipython-input-3-513c0ed21cb9>:286: SettingWithCopyWarning: \n",
            "A value is trying to be set on a copy of a slice from a DataFrame.\n",
            "Try using .loc[row_indexer,col_indexer] = value instead\n",
            "\n",
            "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
            "  df_gene_coord['hic_node_id'] = df_gene_coord['hic_node_id'].astype('int32', copy=False)\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "chr13_H3K27me3.count\n",
            "chr13_H3K36me3.count\n",
            "chr13_H3K4me1.count\n",
            "chr13_H3K4me3.count\n",
            "chr13_H3K9me3.count\n",
            "Test\n",
            "[(9609, 1), (9609, 1), (9609, 1), (9609, 1), (9609, 1)]\n",
            "(32084530,)\n",
            "(32084530,)\n",
            "9609\n",
            "h\n",
            "/content/drive/Shareddrives/DeepBio/GC Merge Data/data/E116/10000/chr13/hic_sparse_redone_h3k27ac_addition.npz\n",
            "c\n",
            "r\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "pDtx_zDFXk7J"
      }
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "provenance": [],
      "gpuType": "T4",
      "machine_shape": "hm"
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}